qpmodel/RulesTrans.cs

```diff
@@ -197,4 +200,125 @@ public override CGroupMember Apply(CGroupMember expr)
             return new CGroupMember(ab_c, expr.group_);
         }
     }
+
+    public class AggSplitRule : ExplorationRule
+    {
+        public override bool Appliable(CGroupMember expr)
+        {
+            LogicAgg agg = expr.logic_ as LogicAgg;
+            if (agg is null || agg.is_local_ || agg.isDerived_)
+                return false;
+
+            return true;
+        }
+
+        public override CGroupMember Apply(CGroupMember expr)
+        {
+            // for manually binding the expression
+            void manualbindexpr(Expr e)
+            {
+                e.bounded_ = true;
+                e.type_ = new BoolType();
+            }
+
+            // for transformaing original having according to the new agg func
+            BinExpr processhaving(Expr e, Dictionary<Expr, Expr> dict)
+            {
+                var be = e as BinExpr;
+                Debug.Assert(be != null);
+                bool isreplace = false;
+                List<Expr> children = new List<Expr>();
+                foreach (var child in be.children_)
+                {
+                    if (dict.ContainsKey(child))
+                    {
+                        children.Add(dict[child]);
+                        isreplace = true;
+                    }
+                    else
+                        children.Add(child);
+                }
+                Debug.Assert(isreplace);
+                return new BinExpr(children[0], children[1], be.op_);
+            }
+
+            LogicAgg origAggNode = (expr.logic_ as LogicAgg);
+            var childNode = (origAggNode.child_() as LogicMemoRef).Deref<LogicNode>();
+
+            var groupby = origAggNode.groupby_?.CloneList();
+            var having = origAggNode.having_?.Clone();
+
+            // process the aggregation functions
+            origAggNode.GenerateAggrFns(false);
+
+            List<AggFunc> aggfns = new List<AggFunc>();
+            origAggNode.aggrFns_.ForEach(x => aggfns.Add(x.Clone() as AggFunc));
+            // need to make aggrFns_ back to null list
+            origAggNode.aggrFns_ = new List<AggFunc>();
+
+            var globalfns = new List<Expr>();
+            var localfns = new List<Expr>();
+
+            // record the processed aggregate functions
+            var derivedAggFuncDict = new Dictionary<Expr, Expr>();
+
+            foreach (var func in aggfns)
+            {
+                Expr processed = null;
+                // average is transformed into sum(sum()) / sum(count())
+                if (func is AggAvg)
```

Comment:
is it possible to move these logics sitting together with corresponding aggfns?

Reply:
I am not sure I understood what you are asking about, at least, not right away. I got a general feel for the aggregate split though.

Reply:
> is it possible to move these logics sitting together with corresponding aggfns?

That should be doable, I will move that.